src/net/transport.rs

```diff
@@ -30,115 +36,134 @@ impl HttpTransport {
+    pub async fn from_tcp(&self, _socket: TcpListener) {
+        /*
         tracing::info!("starting server");
         self.server.bind(BindTarget::HttpListener(socket)).await;
+        */
+        todo!(); // TODO(server)
     }
 
     /// Binds self to port described in `peers_conf`.
     /// # Panics
     /// if self id not found in `peers_conf`
+    #[allow(clippy::unused_async)] // TODO: temporary
     pub async fn bind(&self) -> (SocketAddr, JoinHandle<()>) {
+        /*
         let this_conf = &self.conf.peers()[self.id];
         let port = this_conf.origin.port().unwrap();
         let target = BindTarget::Http(format!("0.0.0.0:{}", port.as_str()).parse().unwrap());
         tracing::info!("starting server; binding to port {}", port.as_str());
         self.server.bind(target).await
+        */
+        todo!(); // TODO(server)
     }
 }
 
 #[async_trait]
 impl Transport for Arc<HttpTransport> {
-    type CommandStream = ReceiverStream<CommandEnvelope>;
+    type RecordsStream = Empty<Vec<u8>>; // TODO(server): resolve placeholder
 
     fn identity(&self) -> HelperIdentity {
-        self.id
+        self.identity
     }
 
-    async fn subscribe(&self, subscription: SubscriptionType) -> Self::CommandStream {
-        match subscription {
-            SubscriptionType::QueryManagement => ReceiverStream::new(
-                self.subscribe_receiver
-                    .lock()
-                    .unwrap()
-                    .take()
-                    .expect("subscribe should only be called once"),
-            ),
-            SubscriptionType::Query(query_id) => {
-                let mut ongoing_queries = self.ongoing_queries.lock().unwrap();
-                match ongoing_queries.entry(query_id) {
-                    Entry::Occupied(_) => {
-                        panic!("attempted to subscribe to commands for query id {}, but there is already a previous subscriber", query_id.as_ref())
-                    }
-                    Entry::Vacant(entry) => {
-                        let (tx, rx) = mpsc::channel(1);
-                        entry.insert(tx);
-                        ReceiverStream::new(rx)
-                    }
-                }
+    async fn send<
+        D: Stream<Item = Vec<u8>> + Send + 'static,
+        Q: QueryIdBinding,
+        S: StepBinding,
+        R: RouteParams<RouteId, Q, S>,
+    >(
+        &self,
+        dest: HelperIdentity,
+        route: R,
+        data: D,
+    ) -> Result<(), TransportError>
+    where
+        Option<QueryId>: From<Q>,
+        Option<Step>: From<S>,
+    {
+        match route.resource_identifier() {
+            RouteId::Records => {
+                let resp_future = self.clients[dest]
+                    .step(
+                        dest,
+                        <Option<QueryId>>::from(route.query_id()).expect("TODO"),
+                        <Option<Step>>::from(route.step()).as_ref().expect("TODO"),
```

Comment:
```suggestion
                        <Option<QueryId>>::from(route.query_id()).ok_or(Error::Something)?,
                        <Option<Step>>::from(route.step()).as_ref().ok_or(Error::Something)?,
```
Right?

Reply:
Certainly `expect("TODO")` is not a keeper. I do think this could be written as an assertion -- this is not processing external requests, it's called by our own code, and we shouldn't write code that tries to send records without an associated query or step ID. A better possibility is changing the API to use an enum like [`TransportCommand`](https://github.com/andyleiserson/ipa/blob/aeaf2cc25c07d63cf3acc0daeab6be570c74cab2/src/helpers/transport/mod.rs#L22) so we can infallibly extract fields that are required for a given request type.

Reply:
Or another possibility is splitting `Transport::send` into multiple functions (`send_record_stream`, `prepare_query`, etc.). If the expectation (or not) of an immediate acknowledgement is implied by the request type, using multiple functions could help keep track of that.